File: scripts/Combination Sum IV/Combination Sum IV.java
class Solution {
    public int combinationSum4(int[] nums, int target) {
        Integer[] memo = new Integer[target + 1];
        return recurse(nums, target, memo);
    }
    
    public int recurse(int[] nums, int remain, Integer[] memo){
        
        if(remain < 0) return 0;
        if(memo[remain] != null) return memo[remain];
        if(remain == 0) return 1;
        
        int ans = 0;
        for(int i = 0; i < nums.length; i++){
            ans += recurse(nums, remain - nums[i], memo);
        }
        
        memo[remain] = ans;
        return memo[remain];
    }
}

File: scripts/Combination Sum IV/Combination Sum IV.js
var combinationSum4 = function(nums, target) {
    const dp = Array(target + 1).fill(0);
    
    nums.sort((a,b) => a - b);
    
    for(let k=1; k <= target; k++) {
        for(let n of nums) {
            if(k < n) break;
            dp[k] += (k == n) ? 1 : dp[k-n];
        }
    }
     
    return dp[target];
};
